Packages/com.unity.inputsystem/InputSystem/Plugins/InputForUI/InputSystemProvider.cs

```diff
@@ -0,0 +1,679 @@
+            {
+                ResetSeenEvents();
+                m_ResetSeenEventsOnUpdate = false;
+            }
+
+            m_Events.Clear();
+        }
+
+        void ResetSeenEvents()
+        {
+            m_SeenTouchEvents = false;
+            m_SeenPenEvents = false;
+        }
+
+        //TODO: Refactor as there is no need for having almost the same implementation in the IM and ISX?
+        void DirectionNavigation(DiscreteTime currentTime)
+        {
+            var(move, axesButtonWerePressed) = ReadCurrentNavigationMoveVector();
+            var direction = NavigationEvent.DetermineMoveDirection(move);
+
+            // Checks for next/previous directions if no movement was detected
+            if (direction == NavigationEvent.Direction.None)
+            {
+                direction = ReadNextPreviousDirection();
+                axesButtonWerePressed = m_NextPreviousAction.WasPressedThisFrame();
+            }
+
+            if (direction == NavigationEvent.Direction.None)
+            {
+                m_RepeatHelper.Reset();
+            }
+            else
+            {
+                if (m_RepeatHelper.ShouldSendMoveEvent(currentTime, direction, axesButtonWerePressed))
+                {
+                    EventProvider.Dispatch(Event.From(new NavigationEvent
+                    {
+                        type = NavigationEvent.Type.Move,
+                        direction = direction,
+                        timestamp = currentTime,
+                        eventSource = GetEventSource(GetActiveDeviceFromDirection(direction)),
+                        playerId = k_DefaultPlayerId,
+                        eventModifiers = m_EventModifiers
+                    }));
+                }
+            }
+        }
+
+        InputDevice GetActiveDeviceFromDirection(NavigationEvent.Direction direction)
+        {
+            switch (direction)
+            {
+                case NavigationEvent.Direction.Left:
+                case NavigationEvent.Direction.Up:
+                case NavigationEvent.Direction.Right:
+                case NavigationEvent.Direction.Down:
+                    return m_MoveAction.action.activeControl.device;
+                case NavigationEvent.Direction.Next:
+                case NavigationEvent.Direction.Previous:
+                    return m_NextPreviousAction.activeControl.device;
+                case NavigationEvent.Direction.None:
+                default:
+                    return Keyboard.current;
+            }
+        }
+
+        (Vector2, bool) ReadCurrentNavigationMoveVector()
+        {
+            var move = m_MoveAction.action.ReadValue<Vector2>();
+            // Check if the action was "pressed" this frame to deal with repeating events
+            var axisWasPressed = m_MoveAction.action.WasPressedThisFrame();
+            return (move, axisWasPressed);
+        }
+
+        NavigationEvent.Direction ReadNextPreviousDirection()
```

Comment:
Have you considered this returning a tuple like the prior function to avoid the use of m_NextPreviousAction directly in the DirectionNavigation function above
axesButtonWerePressed = m_NextPreviousAction.WasPressedThisFrame();

Not critical for this PR